### ISSUE TITLE
Confusion about the Table1 "Quantitative comparison of shape reconstruction performance by Chamfer Distance"

### ISSUE THREAD

**Issue by ifrozenwhale** [SilenKZYoung/CuboidAbstractionViaSeg#5]
Hello, I encountered a little problem when trying to reproduce the reconstruction performance indicators in your paper. Since there is no such part in the project code, I ask a question here and hope to get your reply. Thank you very much!
Based on the pre-trained model, I uniformly sampled the generated geometry according to the method of the paper to obtain a predicted point cloud of 4096 points, and calculate the Chamfer Distance between it and the original point cloud (I used the pcu library to Calculation,[ link ](https://github.com/fwilliams/point-cloud-utils).)
The result obtained is roughly an order of magnitude away from yours (**0.04** (in the paper) vs **0.004** (my experiment result)).
I Note that the point cloud is scale-transformed，as mentioned in the paper, "All shapes are pre-aligned and normalized to the unit scale", I am not sure if there are some other operations before calculating the CD distance.
What should I do to get the CD index in the paper? Thanks very much!

Attachment: The left side of the figure is my result after uniform sampling, and the right side is the original point cloud result. I calculated the CD between these two point clouds.![image-20211129221534576](https://user-images.githubusercontent.com/48271738/143982080-342a46b4-802a-4cf2-b986-ee44c77620f1.png)

**Comment by SilenKZYoung** [comment]
In fact, for each point, we calculate the square of the distance to the nearest point in the other point set and average the distance square of all points. After calculating the average of the bidirectional distance square for the whole dataset, we time the final average distance square by 1000 for a clearer showing.

**Comment by ifrozenwhale** [comment]
Thanks very much! For each point cloud, I think it is calculated according to the following formula. Is my understanding correct?
![image-20211130141501612](https://user-images.githubusercontent.com/48271738/143995453-9bc8cb1c-d875-44ec-af5d-543a2c8ef2ca.png)

**Comment by ifrozenwhale** [comment]
Hi, sorry to disturb you again! I have another question about **sampling in geometry**. Currently, I sample as follows:
For each cube that composes the geometry (assuming a total of M cubes):
1. Move the cube to the origin of the coordinates
2. Calculate the rotation matrix of its rotation to the standard posture
3. At the origin of the coordinates, randomly and uniformly sample according to the scale of the cube (including the left and right boundaries of the scale), and sample N points
4. Rotate the sampled N points to the original position according to the previous rotation matrix
5. Move the sampling point to the original position
Finally, it is allocated according to the volume of each cube, and 4096 points are sampled on the entire point cloud (N*M points).

In this way, sampling on the geometry generated by the pre-training model and calculating the L2 CD index still has a certain gap with the effect in the paper.

- table_test: 1.046(my) vs 0.848(paper)
- chair_test: 0.927(my) vs 0.399(paper)
- airplane_test: 0.466(my) vs 0.329(paper)

I am not sure if this gap is caused by my sampling method.

**Comment by SilenKZYoung** [comment]
> Thanks very much! For each point cloud, I think it is calculated according to the following formula. Is my understanding correct? ![image-20211130141501612](https://user-images.githubusercontent.com/48271738/143995453-9bc8cb1c-d875-44ec-af5d-543a2c8ef2ca.png)
Actually, we are averaging the distances in both directions, not summing them, so we should get a distance value that is half of the one obtained by your formula

**Comment by SilenKZYoung** [comment]
> Thanks very much! For each point cloud, I think it is calculated according to the following formula. Is my understanding correct? ![image-20211130141501612](https://user-images.githubusercontent.com/48271738/143995453-9bc8cb1c-d875-44ec-af5d-543a2c8ef2ca.png)


Actually, we are averaging the distances in both directions, not summing them, so we should get a distance value that is half of the one obtained by your formula

**Comment by SilenKZYoung** [comment]
> 



> 

The sampling manner does have some influence on the final evaluation value, we use sample_points_uniformly in open3d to directly sample the mesh to get a uniform point cloud.

**Comment by ifrozenwhale** [comment]
> > 
> 
> > 
> 
> The sampling manner does have some influence on the final evaluation value, we use sample_points_uniformly in open3d to directly sample the mesh to get a uniform point cloud.

Well, I will try it. Please allow me to ask one more question, is the final result evaluated using cube_masked? Or cube_masked_all_pred? Thanks for your reply, most of my doubts were answered!

**Comment by andrehfr** [comment]
Hi @ifrozenwhale, @SilenKZYoung,
I am currently working on my bachelor thesis at RWTH Aachen University. For the thesis I also need evaluation code for the reconstruction performance of this network to compare it against my approach. Is it possible for one of you to provide the code? This would be really helpful.
